_posts/2020-09-06-how-the-system-tests-works.md

```diff
@@ -0,0 +1,493 @@
+// additional changes...
+kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
+    .withNewMetadata()
+        .withName(topicName)
+        .withNamespace(ResourceManager.kubeClient().getNamespace())
+        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
+    .endMetadata()
+    .editSpec()
+        .withPartitions(partitions)
+        .withReplicas(replicas)
+        .addToConfig("min.insync.replicas", minIsr)
+    .endSpec();
+```
+
+Everything is built within [three stacks](https://github.com/strimzi/strimzi-kafka-operator/blob/master/systemtest/src/main/java/io/strimzi/systemtest/resources/ResourceManager.java#L77-L79). 
+These three stacks are managed by `ResourceManager`, which is taking care of switching. 
+First one's, method stack, responsibility is to hold all resources, which are invoked in the test cases. 
+Second one's, the class stack, responsibility is to hold all resources, which are invoked in the `@BeforeAll` notation. 
+The third stack is used as a pointer to either class or method stack to delete proper resources in specific test phases.
+The pointer stack points to the class stack in we are in the `@BeforeAll` scope, otherwise we are working with the method stack. 
+The logic inside these stacks are that once you create some resource (for instance KafkaTopic),
+it will be pushed inside the stack and the deletion of topic will be performed in the teardown phase. 
+Once the teardown phase is initiated, the method stack will pop all resources after test case is over. 
+The same logic applies with class stack, which will pop all resources when test suite is over. 
+What is worth to mention, is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 
+phase and all created resources will remain in the stack. Also, mostly great for debugging because it could break tests 
+if it's set for multiple than one.
+
+## <a id="auxiliaryclasses">Auxiliary classes</a>
+
+In `system test` module there are many helper classes,  each having a different purpose:
+
+1. Utils
+2. Constants
+3. Environment
+4. Kubernetes client
+5. Kafka clients
+
+#### Utils
+
+These classes are applicable on many problems. We have separate utils for each resource. 
+For instance we have KafkaUtils for KafkaResource, KafkaTopic for KafkaResource and so on. Furthermore, we have got
+Kubernetes related utils such as ServiceUtils, PodUtils, DeploymentsUtils and many more.
+These classes come handy for instance when you have to wait for some change, make some API call, create, change or delete a Kubernetes resource and much more.
+The one method which stands out is `waitFor`, which is used in many utils methods.
+You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert
+that this value was correctly updated. This is a very good example where `race condition` comes into play. This race condition
```

Comment:
I'm not sure _race condition_ is the right term here. I would maybe more explain how the things happen asynchronously and with rolling updates etc. can take seconds or minutes to happen and that is why the `waitFor` is important and useful.